File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void f(int index,vector<int>& candidates, int target,vector<int> &temp,vector<vector<int>> &ans){
        if(target == 0){
            ans.push_back(temp);
            return;
        }

        if(index >= candidates.size() || target < 0){
            return;
        }


        //Include
        temp.push_back(candidates[index]);
        target -= candidates[index];
        f(index,candidates,target,temp,ans);

        //Backtrack
        temp.pop_back();
        target += candidates[index];

        //Exclude
        f(index + 1,candidates,target,temp,ans);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> temp;

        f(0,candidates,target,temp,ans);

        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public void f(int index,int[] candidates, int target,List<Integer> ls,List<List<Integer>> ans){
        if(target == 0){
            ans.add(new ArrayList<>(ls));
            return;
        }
        if(index >= candidates.length || target < 0){
            return;
        }

        //Include

        ls.add(candidates[index]);
        target -= candidates[index];
        f(index,candidates,target,ls,ans);

        //Backtrack
        ls.remove(ls.size() - 1);
        target += candidates[index];

        //Exclude
        f(index + 1,candidates,target,ls,ans);

    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        f(0,candidates,target,new ArrayList<>(),ans);

        return ans;       
    }
}